### PR TITLE
Ethereum tuple decoding

### LINKED ISSUE #1628: [Ethereum] Support decoding `tuple` type
https://etherscan.io/address/0x11111112542d85b3ef69ae05771c2dccff4faa26#code

The second parameter in 1inch router `swap` method is `tuple`:

```json
"7c025200": {
    "inputs": [
      { "name": "caller", "type": "address" },
      {
        "components": [
          { "name": "srcToken", "type": "address" },
          { "name": "dstToken", "type": "address" },
          { "name": "srcReceiver", "type": "address" },
          { "name": "dstReceiver", "type": "address" },
          { "name": "amount", "type": "uint256" },
          { "name": "minReturnAmount", "type": "uint256" },
          { "name": "flags", "type": "uint256" },
          { "name": "permit", "type": "bytes" }
        ],
        "name": "desc",
        "type": "tuple"
      },
      { "name": "data", "type": "bytes" }
    ],
    "name": "swap",
    "outputs": [
      { "name": "returnAmount", "type": "uint256" },
      { "name": "gasLeft", "type": "uint256" }
    ],
    "stateMutability": "payable",
    "type": "function"
  }
```

https://etherscan.io/tx/0xc2d113151124579c21332d4cc6ab2b7f61e81d62392ed8596174513cb47e35ba

Data: 
```bash
0x7c02520000000000000000000000000027239549dd40e1d60f5b80b0c4196923745b1fd2000000000000000000000000000000000000000000000000000000000000006000000000000000000000000000000000000000000000000000000000000001800000000000000000000000002b591e99afe9f32eaa6214f7b7629768c40eeb39000000000000000000000000a0b86991c6218b36c1d19d4a2e9eb0ce3606eb4800000000000000000000000027239549dd40e1d60f5b80b0c4196923745b1fd20000000000000000000000001611c227725c5e420ef058275ae772b41775e261000000000000000000000000000000000000000000000000000005d0fadb1c0000000000000000000000000000000000000000000000000000000005c31df1da000000000000000000000000000000000000000000000000000000000000000400000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000022000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000002080000000000000000000000069d91b94f0aaf8e8a2586909fa77a5c2c89818d50000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000800000000000000000000000000000000000000000000000000000000000000104128acb080000000000000000000000001611c227725c5e420ef058275ae772b41775e2610000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000005d0fadb1c0000000000000000000000000000000000000000000000000000000001000276a400000000000000000000000000000000000000000000000000000000000000a000000000000000000000000000000000000000000000000000000000000000400000000000000000000000002b591e99afe9f32eaa6214f7b7629768c40eeb39000000000000000000000000a0b86991c6218b36c1d19d4a2e9eb0ce3606eb4800000000000000000000000000000000000000000000000000000000
```

should decode as:

```bash
Function: swap(address, (address,address,address,address,uint256,uint256,uint256,bytes), bytes)
#	Name	Type	Data
1	caller	address	0x27239549DD40E1D60F5B80B0C4196923745B1FD2
1	desc.srcToken	address	0x2b591e99afE9f32eAA6214f7B7629768c40Eeb39
1	desc.dstToken	address	0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48
1	desc.srcReceiver	address	0x27239549DD40E1D60F5B80B0C4196923745B1FD2
1	desc.dstReceiver	address	0x1611C227725c5E420Ef058275AE772b41775e261
1	desc.amount	uint256	6395120000000
1	desc.minReturnAmount	uint256	24748356058
1	desc.flags	uint256	4
1	desc.permit	bytes	0x
3	data	bytes	0x00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000002080000000000000000000000069d91b94f0aaf8e8a2586909fa77a5c2c89818d50000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000800000000000000000000000000000000000000000000000000000000000000104128acb080000000000000000000000001611c227725c5e420ef058275ae772b41775e2610000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000005d0fadb1c0000000000000000000000000000000000000000000000000000000001000276a400000000000000000000000000000000000000000000000000000000000000a000000000000000000000000000000000000000000000000000000000000000400000000000000000000000002b591e99afe9f32eaa6214f7b7629768c40eeb39000000000000000000000000a0b86991c6218b36c1d19d4a2e9eb0ce3606eb4800000000000000000000000000000000000000000000000000000000
```

### PR DESCRIPTION
**Pull request by catenocrypt**
## Description

Add support for tuples in Eth function decoding.  Fixes #1628 .

## Testing instructions

Unit tests, hints: Abi, ContractCall.  https://etherscan.io/tx/0xc2d113151124579c21332d4cc6ab2b7f61e81d62392ed8596174513cb47e35ba

## Types of changes

* New feature (non-breaking change which adds functionality)

## Checklist

- [x] Prefix PR title with `[WIP]` if necessary.
- [x] Add tests to cover changes as needed.
- [x] Update documentation as needed.
- [x] I have read the [guidelines](https://developer.trustwallet.com/wallet-core/newblockchain#integration-criteria) for adding a new blockchain
- [x] If there is a related Issue, mention it in the description (e.g. Fixes #<issue_number> ).

### CODE REVIEW

File: src/Ethereum/ABI/Tuple.cpp [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+using namespace TW;
+using namespace TW::Ethereum::ABI;
+
+
+size_t ParamTuple::getSize() const {
+    return _params.getSize();
+}
+
+bool ParamTuple::setValueJson(const std::string& value) {
+    // TODO
```
Comment (hewigovens) [inline src/Ethereum/ABI/Tuple.cpp:22]: what is `setValueJson` for?

File: src/Ethereum/ABI/Tuple.h [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        if (!_isDynamic && param->isDynamic()) {
+            _isDynamic = true;
+        }
+        return _params.addParam(param);
+    }
+    /// Get a parameter.
+    bool getParam(int paramIndex, std::shared_ptr<ParamBase>& param_out) {
+        return _params.getParam(paramIndex, param_out);
+    }
+    /// Return the function type signature, of the form "baz(int32,uint256)"
```
Comment (hewigovens) [inline src/Ethereum/ABI/Tuple.h:36]:
```suggestion
    /// Return the type signature, of the form "(int32,uint256)"
```

**Review by hewigovens — APPROVED** [review]
🚀

File: src/Ethereum/ABI/Tuple.cpp (hunk 1 above)
Comment (catenocrypt) [reply src/Ethereum/ABI/Tuple.cpp]: It's used by Eip712

**Review by vikmeup — APPROVED** [review]

File: tests/Ethereum/ContractCallTests.cpp [hunk 3]
```diff
@@ -180,8 +180,43 @@ TEST(ContractCall, GetAmountsOut) {
...
...":"bytes","value":"0x00000000"}],"name":"desc","type":"tuple"},{"name":"data","type":"bytes","value":"0x00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000002080000000000000000000000069d91b94f0aaf8e8a2586909fa77a5c2c89818d50000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000800000000000000000000000000000000000000000000000000000000000000104128acb080000000000000000000000001611c227725c5e420ef058275ae772b41775e2610000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000005d0fadb1c0000000000000000000000000000000000000000000000000000000001000276a400000000000000000000000000000000000000000000000000000000000000a000000000000000000000000000000000000000000000000000000000000000400000000000000000000000002b591e99afe9f32eaa6214f7b7629768c40eeb39000000000000000000000000a0b86991c6218b36c1d19d4a2e9eb0ce3606eb4800000000000000000000000000000000000000000000000000000000"}]})|";
```
Comment (hewigovens) [inline tests/Ethereum/ContractCallTests.cpp:200]: @catenocrypt could we compare formatted json here? easier to read

**Pull request merged**